ada_feeding_actionselect/ada_feeding_actionselect/policy_service.py

```diff
@@ -0,0 +1,275 @@
+        assert issubclass(
+            context_cls, ContextAdapter
+        ), f"{context_cls_name} must subclass ContextAdapter"
+
+        context_kwargs = self.get_kwargs("context_kws", "context_kwargs")
+        self.context_adapter = context_cls(**context_kwargs)
+
+        # Get the posthoc adapter
+        posthoc_cls_param = self.declare_parameter(
+            "posthoc_class",
+            descriptor=ParameterDescriptor(
+                name="posthoc_class",
+                type=ParameterType.PARAMETER_STRING,
+                description=(
+                    "The class of the posthoc adapter to run, must subclass "
+                    "PosthocAdapter. E.g., ada_feeding_actionselect.adapters.NoContext"
+                ),
+                read_only=True,
+            ),
+        )
+
+        posthoc_cls_name = posthoc_cls_param.value
+        if posthoc_cls_name is None:
+            raise ValueError("Posthoc Adapter is required.")
+
+        posthoc_cls = import_from_string(posthoc_cls_name)
+        assert issubclass(
+            posthoc_cls, PosthocAdapter
+        ), f"{posthoc_cls_name} must subclass PosthocAdapter"
+
+        posthoc_kwargs = self.get_kwargs("posthoc_kws", "posthoc_kwargs")
+        self.posthoc_adapter = context_cls(**posthoc_kwargs)
+
+        # Initialize and validate the policy
+        self.policy = policy_cls(
+            self.context_adapter.dim, self.posthoc_adapter.dim, **kwargs
+        )
+
+        # Start image subscribers
+        self.image = None
+        self.depth = None
+        self.ros_objs = []
+
+        if self.context_adapter.need_rgb:
+            self.ros_objs.append(self.create_subscription(
+                get_img_msg_type("~/image", self), "~/image", self.rgb_callback, 1
+            ))
+        if self.context_adapter.need_depth:
+            self.ros_objs.append(self.create_subscription(
+                get_img_msg_type("~/depth", self), "~/depth", self.depth_callback, 1
+            ))
+
+        self.ros_objs.append(self.create_service(
+            AcquisitionSelect, "~/action_select", self.select_callback
+        ))
+        self.ros_objs.append(self.create_service(
+            AcquisitionReport, "~/action_report", self.report_callback
+        ))
+
+        self.get_logger().info(f"Policy '{policy_name}' initialized!")
+
+    # Subscriptions
+    def rgb_callback(self, msg: Union[CompressedImage, Image]):
+        """
+        Store rgb image
+        """
+        self.image = ros_msg_to_cv2_image(msg)
+
+    def depth_callback(self, msg: Union[CompressedImage, Image]):
+        """
+        Store depth image
+        """
+        self.depth = ros_msg_to_cv2_image(msg)
+
+    # Services
+    def select_callback(self, request, response):
+        """
+        Implement AcquisitionSelect.srv
+        """
+
+        # Run the context adapter
+        context = self.context_adapter.get_context(
+            request.food_context, self.image, self.depth
```

Comment:
In general, the services and subscribers can be called from separate threads. Therefore, `self.image` and `self.depth` should be protected with `threading.Lock` instances. See [food detection](https://github.com/personalrobotics/ada_feeding/blob/60af2be5fdce0cbb7cab5897820cef28bf96d9a7/ada_feeding_perception/ada_feeding_perception/segment_from_point.py#L278) as an example.